pkg/storage/mock_helpers_test.go

```diff
@@ -0,0 +1,40 @@
+package storage_test
+
+import (
+	"fmt"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+)
+
+type mockSequence struct {
+	first, last *gomock.Call // may be both nil (= the default value of mockSequence) to mean empty sequence
+}
+
+// like gomock.inOrder, but can be nested
+func inOrder(calls ...interface{}) mockSequence {
```

Comment:
This might be a valuable addition to gomock as well, do you think we should port it there?

Reply:
Hum… I did spend some time looking in GoMock for various helpers, including ultimately something like this, so maybe it would be natural to propose.

OTOH (and this is with a huge disclaimer: this is the first project I’ve encountered with so much state to mock, and significant use of mocks in general, and I have not tried testing it in any other way at all, so I am really not justified to comment on the trade-offs or suitability in any way), it seems to me that once the use of mocking gets so widespread like here, that a several-level call stack of setting up mocks turns out to be necessary to make the mocks readable, it might be an indication that things have gone too far down a difficult road, and the tests might need reconsidering / restructuring in some other way (again, disclaimer: I have no idea what that would be) instead of making it any easier to go down that path.

(From what I’ve been able to gather, the GoMock implementation is not set up for enforcing a strict _order_ of calls, it seems to be more structured about ~stateless pairs of input→output that could be called in any order.  Which is one reason why test failures about the code call sequence not matching the mock sequence are so convoluted and difficult to read, they almost always refer to some other distant mocked call of the same method with different arguments and only secondarily mention the ordering constraints.)